File: src/dotnet/src/Dependabot.Job/Models/DependencyGroup.cs
using System.Collections.Immutable;

namespace Dependabot.Job;

/// <summary>
/// Represents a dependency group configuration.
/// </summary>
public sealed record DependencyGroup
{
  /// <summary>
  /// Gets or sets the name of the dependency group.
  /// </summary>
  public required string Name { get; init; }

  /// <summary>
  /// Gets or sets what this group applies to (e.g., "version-updates", "security-updates").
  /// </summary>
  public string? AppliesTo { get; init; }

  /// <summary>
  /// Gets or sets the dependency type.
  /// </summary>
  public string? DependencyType { get; init; }

  /// <summary>
  /// Gets or sets the update types allowed for this group.
  /// </summary>
  public ImmutableArray<string> UpdateTypes { get; init; } = [];

  /// <summary>
  /// Gets or sets the patterns for dependencies to include.
  /// </summary>
  public ImmutableArray<string> Patterns { get; init; } = [];

  /// <summary>
  /// Gets or sets the patterns for dependencies to exclude.
  /// </summary>
  public ImmutableArray<string> ExcludePatterns { get; init; } = [];
}

File: src/ruby/lib/dependabot/job/models/dependency_group.rb
# typed: strict
# frozen_string_literal: true

require 'sorbet-runtime'

module Dependabot
  module Job
    module Models
      # Represents a dependency group configuration.
      class DependencyGroup < T::Struct
        extend T::Sig

        # The name of the dependency group.
        const :name, String

        # What this group applies to (e.g., "version-updates", "security-updates").
        const :applies_to, T.nilable(String), default: nil

        # The dependency type.
        const :dependency_type, T.nilable(String), default: nil

        # The update types allowed for this group.
        const :update_types, T::Array[String], default: []

        # The patterns for dependencies to include.
        const :patterns, T::Array[String], default: []

        # The patterns for dependencies to exclude.
        const :exclude_patterns, T::Array[String], default: []

        sig { params(hash: T::Hash[Symbol, T.untyped]).returns(DependencyGroup) }
        def self.from_hash(hash)
          new(
            name: hash.fetch(:name),
            applies_to: hash[:applies_to],
            dependency_type: hash[:dependency_type],
            update_types: Array(hash[:update_types] || []),
            patterns: Array(hash[:patterns] || []),
            exclude_patterns: Array(hash[:exclude_patterns] || [])
          )
        end

        sig { returns(T::Hash[Symbol, T.untyped]) }
        def to_hash
          {
            name: name,
            applies_to: applies_to,
            dependency_type: dependency_type,
            update_types: update_types,
            patterns: patterns,
            exclude_patterns: exclude_patterns
          }.compact
        end
      end
    end
  end
end
